Runtime.ts

```diff
@@ -0,0 +1,59 @@
+import { Atom } from "./Atom.ts";
+import { EffectLike, isEffectLike } from "./common.ts";
+import { EffectId } from "./Effect.ts";
+import { EffectState } from "./EffectState.ts";
+import { Name } from "./Name.ts";
+
+export class Runtime {
+  #contexts = new WeakMap<object, Context>();
+
+  // TODO: enable hooking into context for rewrites
+  run = <S extends boolean, V, E extends Error, T>(
+    root: EffectLike<S, V, E, T>,
+    env: unknown extends V ? object : V,
+  ): ColoredResult<S, E | T> => {
+    const envSafe = env as object; // for now
+    let context = this.#contexts.get(envSafe);
+    if (!context) {
+      context = new Context(envSafe);
+      this.#contexts.set(envSafe, context);
+    }
```

Comment:
wow, this enables to re-use a previous computation

```
const root = add(1, add(2, add(3, 4)));

const result = run(root, context);

console.log("result", result);

const root2 = add(100, add(2, add(3, 4)));

const result2 = run(root2, context);

console.log("result2 used cached", result2);
```

I believes it depends on `sig.ref` and `sig.val` to generate the same keys.

Reply:
Bingo! 

On a related note, we may eventually want to expose a means of explicitly clearing the map underlying `sig.val`.